webapp/javascript/util/flamebearer.js

```diff
@@ -17,12 +17,67 @@
 // This component is based on flamebearer project
 //   https://github.com/mapbox/flamebearer
 
-export function deltaDiff(levels) {
+function deltaDiff(levels, start, step) {
   for (const level of levels) {
     let prev = 0;
-    for (let i = 0; i < level.length; i += 4) {
+    for (let i = start; i < level.length; i += step) {
       level[i] += prev;
       prev = level[i] + level[i + 1];
     }
   }
 }
+
+export function deltaDiffWrapper(format, levels) {
+  if (format === "double") {
```

Comment:
Can you explain why there are two `deltaDiff` here with different start / steps with one happening right before the other?

Reply:
The corresponding code on the server is here: [diff](https://github.com/pyroscope-io/pyroscope/pull/289/files#diff-b8e94c44499b631822aa06d0e72f48a20addcf5a4cca15e4e55d14745fcf3065R188-R191)

```go
func CombineToFlamebearerStruct(leftTree, rightTree *Tree, maxNodes int) *Flamebearer {
	// ...
	// delta encoding
	deltaEncoding(res.Levels, 0, 7)
	deltaEncoding(res.Levels, 3, 7)
	return &res
}
```

With `format=double`, the response from the server combines values from 2 left and right trees into a single array for `levels`, start at `i+0` and `i+3` with `step=7`:

```
i+0 = x offset, left  tree
i+1 = total   , left  tree
i+2 = self    , left  tree
i+3 = x offset, right tree
i+4 = total   , right tree
i+5 = self    , right tree
i+6 = index in the names array
```

The `levels` array is encoded using delta encoding for x offset on the left and the right trees before being sent to the client. Therefore, the client needs to decode it before rendering.